Gem/Code/Source/Components/ScriptableDecalComponent.cpp

```diff
@@ -0,0 +1,161 @@
+                ->Attribute(AZ::Script::Attributes::Scope, AZ::Script::Attributes::ScopeFlags::Common)
+                ->Attribute(AZ::Script::Attributes::Category, "Rendering")
+                ->Attribute(AZ::Script::Attributes::Module, "rendering")
+                ->Attribute(AZ::Script::Attributes::Scope, AZ::Script::Attributes::ScopeFlags::Common)
+                ->Event(
+                    "SpawnDecal",
+                    &DecalRequestBus::Events::SpawnDecal)
+                ;
+        }
+    }
+
+    void ScriptableDecalComponent::GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& services)
+    {
+        services.push_back(AZ_CRC_CE("ScriptableDecalService"));
+    }
+
+    void ScriptableDecalComponent::GetIncompatibleServices(AZ::ComponentDescriptor::DependencyArrayType& services)
+    {
+        services.push_back(AZ_CRC_CE("ScriptableDecalService"));
+    }
+
+    void ScriptableDecalComponent::Activate()
+    {
+        AZ_Printf("ScriptableDecalComponent", "Activate");
+        m_decalFeatureProcessor = AZ::RPI::Scene::GetFeatureProcessorForEntity<AZ::Render::DecalFeatureProcessorInterface>(GetEntityId());
+        if (m_decalFeatureProcessor)
+        {
+            AZ::RPI::Scene* scene = m_decalFeatureProcessor->GetParentScene();
+            DecalRequestBus::Handler::BusConnect(scene->GetId());
+            AZ::TickBus::Handler::BusConnect();
+        }
+    }
+
+    void ScriptableDecalComponent::Deactivate()
+    {
+        AZ::TickBus::Handler::BusDisconnect();
+        DecalRequestBus::Handler::BusDisconnect();
+        m_decalFeatureProcessor = nullptr;
+    }
+
+    void ScriptableDecalComponent::SpawnDecal(const AZ::Transform& worldTm, AZ::Data::AssetId materialAssetId, const SpawnDecalConfig& config)
+    {
+        DecalHandle handle = m_decalFeatureProcessor->AcquireDecal();
+
+        AZ::Vector3 scale = AZ::Vector3(config.m_scale, config.m_scale, config.m_scale * config.m_thickness);
+
+        m_decalFeatureProcessor->SetDecalTransform(handle, worldTm, scale);
+        m_decalFeatureProcessor->SetDecalMaterial(handle, materialAssetId);
+        m_decalFeatureProcessor->SetDecalOpacity(handle, config.m_opacity);
+        m_decalFeatureProcessor->SetDecalAttenuationAngle(handle, config.m_attenutationAngle);
+        m_decalFeatureProcessor->SetDecalSortKey(handle, config.m_sortKey);
+
+        auto now = AZStd::chrono::steady_clock::now().time_since_epoch();
+        auto nowMs = AZStd::chrono::duration_cast<AZStd::chrono::milliseconds>(now).count();
+        uint32_t currentTimeMs = static_cast<uint32_t>(nowMs);
+
+        if (config.m_lifeTime > 0.0)
+        {
+            uint32_t lifetimeMs = static_cast<uint32_t>(config.m_lifeTime * 1000.0f);
+            uint32_t despawnTimeMs = currentTimeMs + lifetimeMs;
+
+            m_decalHeap.push_back({ config , handle, despawnTimeMs });
+            AZStd::push_heap(m_decalHeap.begin(), m_decalHeap.end(), HeapCompare);
```

Comment:
Code comment on the use/benefit of push_heap. Not for me of course, I _definitely did not_ have to look this up *cough*

Reply:
If you're curious, the priority_queue container wraps up heap functionality quite nicely.. but as the newly added comment makes clear heaps are all about popping off ordered elements as cheaply as possible.